static/js/message_events.js

```diff
@@ -167,6 +167,57 @@ exports.update_messages = function update_messages(events) {
             }
 
             const current_filter = narrow_state.filter();
+            const messages_to_rerender = [];
+            for (const id of event.message_ids) {
+                const msg = message_store.get(id);
+                if (msg === undefined) {
+                    continue;
+                }
+
+                // Remove the recent topics entry for the old topics;
+                // must be called before we call set_message_topic.
+                stream_topic_history.remove_message({
+                    stream_id: msg.stream_id,
+                    topic_name: msg.topic,
+                });
+
+                // Update the unread counts; again, this must be called
+                // before we modify the topic field on the message.
+                unread.update_unread_topics(msg, event);
+
+                // Now edit the attributes of our message object.
+                if (topic_edited) {
+                    msg.topic = new_topic;
+                    msg.topic_links = event.topic_links;
+                }
+                if (stream_changed) {
+                    const new_stream_name = stream_data.get_sub_by_id(new_stream_id).name;
+                    msg.stream_id = event.new_stream_id;
+                    msg.stream = new_stream_name;
+                    msg.display_recipient = new_stream_name;
+                }
+
+                // Add the recent topics entry for the new stream/topics.
+                stream_topic_history.add_message({
+                    stream_id: msg.stream_id,
+                    topic_name: msg.topic,
+                    message_id: msg.id,
+                });
+
+                if (!changed_narrow && current_filter && current_filter.can_apply_locally() &&
```

Comment:
Moving this broke the `!changed_narrow` check, which was an optimization; I think it's correct to just move this around where we use `messages_to_rerender` instead.  I did that.